src/report.rs

```diff
@@ -0,0 +1,404 @@
+
+    fn timestamp(&self) -> u32 {
+        u32::from_le_bytes(self.data[0..4].try_into().unwrap()) // infallible slice-to-array conversion
+    }
+
+    fn breakdown_key(&self) -> BK {
+        BK::deserialize(GenericArray::from_slice(&[self.data[4]]))
+    }
+
+    fn trigger_value(&self) -> u32 {
+        u32::from_le_bytes(self.data[5..9].try_into().unwrap()) // infallible slice-to-array conversion
+    }
+
+    fn encap_key(&self) -> &[u8] {
+        &self.data[9..Self::CIPHERTEXT_OFFSET]
+    }
+
+    fn match_key_ciphertext(&self) -> &[u8] {
+        &self.data[Self::CIPHERTEXT_OFFSET..Self::EVENT_TYPE_OFFSET]
+    }
+
+    fn event_type(&self) -> EventType {
+        EventType::try_from(self.data[Self::EVENT_TYPE_OFFSET]).unwrap() // validated on construction
+    }
+
+    fn key_id(&self) -> KeyIdentifier {
+        self.data[Self::EVENT_TYPE_OFFSET + 1]
+    }
+
+    fn epoch(&self) -> Epoch {
+        u16::from_le_bytes(
+            self.data[Self::EVENT_TYPE_OFFSET + 2..Self::SITE_DOMAIN_OFFSET]
+                .try_into()
+                .unwrap(), // infallible slice-to-array conversion
+        )
+    }
+
+    fn site_domain(&self) -> &str {
+        std::str::from_utf8(&self.data[Self::SITE_DOMAIN_OFFSET..]).unwrap() // validated on construction
+    }
+
+    #[allow(dead_code)] // TODO: temporary
+    fn from_bytes(bytes: &'a [u8]) -> Result<Self, InvalidReportError> {
+        EventType::try_from(bytes[Self::EVENT_TYPE_OFFSET])?;
+        let site_domain = &bytes[Self::SITE_DOMAIN_OFFSET..];
+        if !site_domain.is_ascii() {
+            return Err(NonAsciiStringError::from(site_domain).into());
+        }
+        Ok(Self {
+            data: bytes,
+            phantom_data: PhantomData,
+        })
+    }
+
+    #[allow(dead_code)] // TODO: temporary
+    fn decrypt(&self, key_registry: &KeyRegistry) -> Result<Report<MK, BK>, InvalidReportError> {
+        let info = Info::new(
+            self.key_id(),
+            self.epoch(),
+            self.event_type(),
+            HELPER_ORIGIN,
+            self.site_domain(),
+        )
+        .unwrap(); // validated on construction
+
+        let mut ciphertext: GenericArray<u8, <MK as MatchKeyCrypt>::CiphertextSize> =
+            GenericArray::clone_from_slice(self.match_key_ciphertext());
+        let plaintext = open_in_place(key_registry, self.encap_key(), &mut ciphertext, info)?;
+
+        Ok(Report {
+            timestamp: self.timestamp(),
+            mk_shares: <MK as MatchKeyCrypt>::SemiHonestShares::deserialize(
+                GenericArray::from_slice(plaintext),
+            ),
+            event_type: self.event_type(),
+            breakdown_key: self.breakdown_key(),
+            trigger_value: self.trigger_value(),
+            epoch: self.epoch(),
+            site_domain: self.site_domain().to_owned(),
+        })
+    }
+}
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct Report<MK: MatchKeyCrypt, BK: GaloisField> {
+    pub timestamp: u32,
```

Comment:
Yes. I don't think we have a lot of choice in the matter, because we need the event type in the clear in order to decrypt the match keys. I'd be happy to be wrong about that though.